### PR TITLE
Optimise builder

### PR DESCRIPTION
**Pull request by cristaloleg**
See benchmark diff https://gist.github.com/cristaloleg/619b66359026c5c06f91b55784757cab

### CODE REVIEW

File: algo_es.go [hunk 1]
```diff
@@ -69,6 +69,10 @@ func (h esAlg) Algorithm() Algorithm {
 	return h.alg
 }
 
+func (h esAlg) SignSize() int {
+	return h.privateKey.Curve.Params().BitSize / 8
```
Comment (FZambia) [inline algo_es.go:73]: So sometimes it's useful to divide on 8, right?:)

Comment (cristaloleg) [reply algo_es.go:73]: Yep)

File: algo.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 // Signer is used to sign tokens.
 type Signer interface {
 	Algorithm() Algorithm
+	SignSize() int
```
Comment (cristaloleg) [inline algo.go:11]: Breaking compatibility, but just a little bit.

File: build.go [hunk 3]
```diff
@@ -50,71 +51,115 @@ func (b *Builder) BuildBytes(claims interface{}) ([]byte, error) {
 
 // Build used to create and encode JWT with a provided claims.
```
Comment (cristaloleg) [inline build.go:52]:
```suggestion
// Build used to create and encode JWT with a provided claims.
// If claims param is of type []byte then we treat it as a marshaled JSON.
// In other words you can pass already marshaled claims.
```

File: algo_es.go [hunk 4]
```diff
@@ -69,6 +69,10 @@ func (h esAlg) Algorithm() Algorithm {
 	return h.alg
 }
 
+func (h esAlg) SignSize() int {
+	return (h.privateKey.Curve.Params().BitSize + 7) / 8 * 2
```
Comment (FZambia) [inline algo_es.go:73]: this one seems magical, maybe add comment with formula reasoning?

Comment (cristaloleg) [inline algo_es.go:73]:
```suggestion
	return 2 * h.curveBits
```

File: build.go [hunk 5]
```diff
@@ -49,72 +50,114 @@ func (b *Builder) BuildBytes(claims interface{}) ([]byte, error) {
...
+		case PS512:
+			return encHeaderPS512
+
+		default:
+			// another algorithm? encode below
+		}
+	}
 	// returned err is always nil, see *Header.MarshalJSON
-	buf, _ := header.MarshalJSON()
+	buf, _ := json.Marshal(header)
```
Comment (storozhukBM) [inline build.go:138]: json.Marshal can return error

File: build.go [hunk 6]
```diff
@@ -49,72 +50,114 @@ func (b *Builder) BuildBytes(claims interface{}) ([]byte, error) {
...
 	}
-
-	encoded = make([]byte, base64EncodedLen(len(raw)))
-	base64Encode(encoded, raw)
-
-	return raw, encoded, nil
 }
 
-func encodeHeader(header *Header) []byte {
+func encodeHeader(header *Header) string {
```
Comment (storozhukBM) [inline build.go:99]: Why did you decide to return `string` from this method?

File: build.go (hunk 5 above)
Comment (cristaloleg) [reply build.go:138]: See comment above.

**Pull request merged**